File: js/chart/segments.js
/* segments.js — the market's history as a sequence of EPISODES.
 *
 * A port of sim/tl/segments.py, compared segment-for-segment in
 * tests/test_segment_parity.py.
 *
 * regime.js answers "what is the market doing at bar i". This answers the
 * question an annotated chart answers: "this was a downward channel, then a
 * range, then an upward channel" — a sequence with boundaries. Boundaries are
 * the part a per-bar label cannot give you.
 *
 * Minimum length and hysteresis are what stop it producing confetti: a
 * three-bar excursion is noise inside the episode either side of it, and a
 * regime must hold for `confirmBars` before a new episode opens.
 */

import * as regime from './regime.js';
import { atrSeries } from './tlengine.js';

export const DEFAULT_SEGMENT_PARAMS = {
  minBars: 12,
  confirmBars: 3,
  maxSegments: 12,
};

const LABEL = {
  [regime.TRENDING_UP]: 'Uptrend',
  [regime.TRENDING_DOWN]: 'Downtrend',
  [regime.SIDEWAYS]: 'Range',
  [regime.TRANSITION]: 'Transition',
};

export class Segment {
  constructor(o) { Object.assign(this, o); }
  get label() { return LABEL[this.kind] || this.kind; }
}

/** Raw runs with hysteresis: a new run opens only once `confirm` bars agree. */
function runs(kinds, confirm) {
  const n = kinds.length;
  if (!n) return [];
  const out = [];
  let cur = kinds[0], start = 0, j = 1;
  while (j < n) {
    if (kinds[j] === cur) { j++; continue; }
    let k = j;
    while (k < n && k - j < confirm && kinds[k] === kinds[j]) k++;
    if (k - j >= confirm || k >= n) {
      out.push([cur, start, j - 1]);
      cur = kinds[j];
      start = j;
      j++;
    } else {
      j = k;
    }
  }
  out.push([cur, start, n - 1]);
  return out;
}

/** Merge short runs into the LONGER neighbour — the choice that changes least. */
function absorb(rs, minBars) {
  if (!rs.length) return [];
  let list = rs.map((r) => [...r]);
  let changed = true;
  while (changed && list.length > 1) {
    changed = false;
    for (let idx = 0; idx < list.length; idx++) {
      const [, i0, i1] = list[idx];
      if (i1 - i0 + 1 >= minBars) continue;
      const prevLen = idx > 0 ? list[idx - 1][2] - list[idx - 1][1] : -1;
      const nextLen = idx + 1 < list.length ? list[idx + 1][2] - list[idx + 1][1] : -1;
      if (prevLen < 0 && nextLen < 0) continue;
      if (prevLen >= nextLen) list[idx - 1][2] = i1;
      else list[idx + 1][1] = i0;
      list.splice(idx, 1);
      changed = true;
      break;
    }
  }
  const fused = [list[0]];
  for (const r of list.slice(1)) {
    if (r[0] === fused[fused.length - 1][0]) fused[fused.length - 1][2] = r[2];
    else fused.push(r);
  }
  return fused;
}

/** Episodes over `bars`, oldest first. */
export function build(bars, params = {}) {
  const p = { ...DEFAULT_SEGMENT_PARAMS, ...params };
  const n = bars.length;
  if (!n) return [];
  const r = regime.compute(bars);
  const atr = r.atr || atrSeries(bars, 14);
  const kinds = r.regime.slice(0, n);
  const list = absorb(runs(kinds, p.confirmBars), p.minBars);

  const out = list.map(([kind, i0, i1], k) => {
    let hi = -Infinity, lo = Infinity;
    for (let j = i0; j <= i1; j++) {
      if (bars[j].h > hi) hi = bars[j].h;
      if (bars[j].l < lo) lo = bars[j].l;
    }
    const a0 = Number.isFinite(atr[i0]) ? atr[i0] : 0;
    return new Segment({
      kind, i0, i1, t0: bars[i0].t, t1: bars[i1].t,
      bars: i1 - i0 + 1, high: hi, low: lo,
      retAtr: a0 ? (bars[i1].c - bars[i0].c) / a0 : 0,
      closed: k < list.length - 1,
    });
  });
  return out.slice(-p.maxSegments);
}

File: sim/tl/segments.py
"""
segments.py — the market's history as a sequence of EPISODES, not a per-bar label.

`regime.py` answers "what is the market doing at bar i". Every annotated chart
answers a different question: "this was a downward channel, then a range, then
an upward channel". That is a sequence with boundaries, and boundaries are what
a per-bar label cannot give you -- reading them off by eye means deciding where
one episode ends, which is exactly the judgement this makes explicit.

TWO THINGS STOP IT PRODUCING CONFETTI

  MINIMUM LENGTH. A three-bar excursion into `transition` is not an episode, it
  is noise inside the episode either side of it. Runs shorter than `min_bars`
  are absorbed into their neighbour rather than emitted.

  HYSTERESIS. A regime must hold for `confirm_bars` consecutive bars before a
  new segment opens. Without it every oscillation across a threshold starts a
  new episode, and the segmentation says more about the threshold than about
  the market.

CAUSALITY. Segments are built by a forward walk and each one's start is fixed
when it opens, so `segments_upto(i)` returns what a bar-i observer could have
known. The LAST segment is open-ended and its end moves as bars arrive -- which
is honest: you never know the current episode has ended until it has.
"""

from dataclasses import dataclass
from typing import List

import numpy as np

from . import regime as reg


@dataclass
class Segment:
    """One episode of a single regime."""

    kind: str                 # trending_up | trending_down | sideways | transition
    i0: int
    i1: int
    t0: int
    t1: int
    bars: int
    high: float = 0.0         # extremes reached inside the episode
    low: float = 0.0
    ret_atr: float = 0.0      # close-to-close move, in ATR at the start
    closed: bool = True       # the final segment is still forming

    @property
    def label(self) -> str:
        return {
            reg.TRENDING_UP: 'Uptrend',
            reg.TRENDING_DOWN: 'Downtrend',
            reg.SIDEWAYS: 'Range',
            reg.TRANSITION: 'Transition',
        }.get(self.kind, self.kind)

    def to_row(self) -> dict:
        return {
            'kind': self.kind, 'label': self.label,
            'i0': self.i0, 'i1': self.i1, 't0': self.t0, 't1': self.t1,
            'bars': self.bars, 'high': self.high, 'low': self.low,
            'ret_atr': round(self.ret_atr, 3), 'closed': self.closed,
        }


@dataclass
class SegmentParams:
    min_bars: int = 12        # shorter runs are absorbed, not emitted
    confirm_bars: int = 3     # consecutive bars before a new episode opens
    max_segments: int = 12    # most recent N


def _runs(kinds, confirm):
    """
    Raw runs with hysteresis: a new run opens only once `confirm` consecutive
    bars agree, so a single bar flipping across a threshold does not split an
    episode in two.
    """
    n = len(kinds)
    if not n:
        return []
    out = []
    cur = kinds[0]
    start = 0
    j = 1
    while j < n:
        if kinds[j] == cur:
            j += 1
            continue
        # candidate change: does it hold?
        k = j
        while k < n and k - j < confirm and kinds[k] == kinds[j]:
            k += 1
        if k - j >= confirm or k >= n:
            out.append((cur, start, j - 1))
            cur = kinds[j]
            start = j
            j += 1
        else:
            j = k
    out.append((cur, start, n - 1))
    return out


def _absorb(runs, min_bars):
    """
    Merge runs shorter than `min_bars` into whichever NEIGHBOUR is longer. A
    short run between two different regimes is genuinely ambiguous, and giving
    it to the longer side is the choice that changes the picture least.
    """
    if not runs:
        return []
    runs = [list(r) for r in runs]
    changed = True
    while changed and len(runs) > 1:
        changed = False
        for idx, (kind, i0, i1) in enumerate(runs):
            if i1 - i0 + 1 >= min_bars:
                continue
            prev_len = (runs[idx - 1][2] - runs[idx - 1][1]) if idx > 0 else -1
            next_len = (runs[idx + 1][2] - runs[idx + 1][1]) if idx + 1 < len(runs) else -1
            if prev_len < 0 and next_len < 0:
                continue
            if prev_len >= next_len:
                runs[idx - 1][2] = i1
            else:
                runs[idx + 1][1] = i0
            runs.pop(idx)
            changed = True
            break
    # neighbouring runs of the same kind can now be adjacent: fuse them
    fused = [runs[0]]
    for r in runs[1:]:
        if r[0] == fused[-1][0]:
            fused[-1][2] = r[2]
        else:
            fused.append(r)
    return fused


def build(bars, regimes=None, atr=None, params: SegmentParams = None,
          times=None, upto=None):
    """
    Episodes over `bars`, oldest first. `upto` truncates to a bar-i observer.
    """
    p = params or SegmentParams()
    n = len(bars)
    if not n:
        return []
    i_end = (n - 1) if upto is None else min(upto, n - 1)

    if regimes is None:
        regimes, _, feats = reg.compute(bars)
        atr = feats['atr'] if atr is None else atr
    if atr is None:
        from ..indicators import atr as atr_series
        atr = atr_series(bars, 14)
    if times is None:
        times = np.asarray(bars.index.astype('int64') // 1_000_000)

    high = np.asarray(bars['high'], dtype=float)
    low = np.asarray(bars['low'], dtype=float)
    close = np.asarray(bars['close'], dtype=float)

    kinds = list(regimes[:i_end + 1])
    runs = _absorb(_runs(kinds, p.confirm_bars), p.min_bars)

    out = []
    for k, (kind, i0, i1) in enumerate(runs):
        a0 = atr[i0] if i0 < len(atr) and np.isfinite(atr[i0]) else 0.0
        seg = Segment(
            kind=kind, i0=i0, i1=i1,
            t0=int(times[i0]), t1=int(times[i1]),
            bars=i1 - i0 + 1,
            high=float(np.max(high[i0:i1 + 1])),
            low=float(np.min(low[i0:i1 + 1])),
            ret_atr=float((close[i1] - close[i0]) / a0) if a0 else 0.0,
            closed=(k < len(runs) - 1),
        )
        out.append(seg)
    return out[-p.max_segments:]
